builtin/rebase--helper.c

```diff
@@ -73,18 +79,59 @@ int cmd_rebase__helper(int argc, const char **argv, const char *prefix)
 	flags |= abbreviate_commands ? TODO_LIST_ABBREVIATE_CMDS : 0;
 	flags |= rebase_merges ? TODO_LIST_REBASE_MERGES : 0;
 	flags |= rebase_cousins > 0 ? TODO_LIST_REBASE_COUSINS : 0;
-	flags |= command == SHORTEN_OIDS ? TODO_LIST_SHORTEN_IDS : 0;
 
 	if (rebase_cousins >= 0 && !rebase_merges)
 		warning(_("--[no-]rebase-cousins has no effect without "
 			  "--rebase-merges"));
 
+	if (command == MAKE_SCRIPT && !upstream && squash_onto) {
+		FILE *squash_onto_file;
+		int ret;
+
+		squash_onto_file = fopen(path_squash_onto(), "w");
+		if (squash_onto_file == NULL)
+			return error_errno(_("could not open '%s'"), path_squash_onto());
+
+		ret = fputs(squash_onto, squash_onto_file);
+		fclose(squash_onto_file);
```

Comment:
This loses the newline (not sure it matters, but it is a change in behavior, and I don't think we have to). Also: Don't we already have helpers for this?

*clicketyclick*

`write_file(path, fmt, ...)`. So the code would be this short:

```c
if (command == MAKE_SCRIPT && !upstream && squash_onto &&
    write_file(path_squash_onto(), "%s\n", squash_onto) < 0)
        return error_errno(_("could not write '%s'"), path_squash_onto());
```

Reply:
write_file() does not return an error code. I’m changing this accordingly.

Reply:
Yes, you're right. @peff changed this a while ago. In the exact wrong direction. ☹️

Reply:
To be fair, the return code was a red herring; it would always die anyway. You can be mad at me for removing the `_gently` form, though.

Reply:
I am not mad at you. I am sad that some of us worked so hard to libify Git, and that this change (undoing some of that effort) was so happily applied.

Maybe in the future, your aims are more aligned with mine, now that you see where it would have been very nice indeed to already have a properly "gentle" function.

This `die()`ing left and right is just no good. It is a really sloppy way to develop code.

Reply:
I'm not sure I disagree with the original removal. The gentle version never had any callers and after 4 years was removed. I'm totally happy for you to re-add it if you have an actual caller, but I have no regrets about removing dead code.

Reply:
In light of all the code churn caused by having to libify everything, carefully combing through code to avoid those pesky `exit()` and `die()` calls (doing this in the `deprecate-grafts` patch series took me well over three times as long as to come up with the original patch series!!!), I think you will agree that a better idea is to keep the gentle version, and to convert the non-gentle callers (because they should most likely have been gentle in the first place, that's the problem your review uncovered, not that the gentle version wasn't necessary: it was, and still is, necessary.)

Reply:
> I think that converting some die() calls into error() calls when no other libifying effort is made could be seen as unnecessary code churn and rejected.

Are you trying to excuse yourself from the task? I remember that the builtin `am` suffered from many a `die()`, and I would not be surprised if it still did.

I know it is hard work, and I understand that you don't want to do it if you do not even see the value of it.

Obviously, it would be much easier to stay sloppy and just call `die()`. And leave the user puzzled what happened here, as the code that calls `die()` is unlikely to have the bigger picture of what the user tried to do.

For example, if your builtin `am` runs as part of a rebase, a regular Git user who is not familiar with Git's code base would probably go "ummm" when seeing something about `am` written in an error message: "What is this 'am' thing? I called a *rebase* for crying out loud!"

So I understand that it is so much easier from the developer's point of view to just forgo proper error handling. Quick and dirty. And make a mess of the user experience.

I encourage you to try to understand this reasoning, so that your own code can become better at this.

Reply:
> It is also very hard to predict a long time in advance if and when we will need to libify a piece of code.

I forgot to point out how ludicrous this statements premise is: you seem to think that it is okay to be sloppy by calling `die()`, as long as you think the code is not deep down in the call chain. Forgetting that every useful piece of code will be reused sooner rather than later. So unless you think that your code is not useful, you should write it in a way that it can be reused easily.